File: app/app/core/grade.py
"""统一分级：severity/metrics/magnitude → {band, tone, zh, en}。

前端 realGrade 的服务端实现，单一真相源。tone ∈ neutral|green|yellow|orange|red。
"""
from __future__ import annotations

import re
from typing import Any

TONES = ("neutral", "green", "yellow", "orange", "red")
_ALERT_ZH = {"green": "绿色", "yellow": "黄色", "orange": "橙色", "red": "红色"}
_ALERT_EN = {"green": "Green", "yellow": "Yellow", "orange": "Orange", "red": "Red"}
_CMA_LEVEL_TONE = {"红": "red", "橙": "orange", "黄": "yellow", "蓝": "green"}


def fmt_num(n: Any) -> str:
    try:
        x = float(n)
    except (TypeError, ValueError):
        return str(n)
    if abs(x) >= 10000:
        return f"{int(round(x)):,}"
    if abs(x) >= 100:
        return f"{x:.0f}"
    r = round(x, 1)
    return str(int(r)) if r == int(r) else f"{r:.1f}"


def cma_level_from_text(text: str | None) -> str | None:
    t = text or ""
    if re.search(r"红色预警|红预警", t):
        return "红"
    if re.search(r"橙色预警|橙预警", t):
        return "橙"
    if re.search(r"黄色预警|黄预警", t):
        return "黄"
    if re.search(r"蓝色预警|蓝预警", t):
        return "蓝"
    return None


def is_cma_alert(source: str | None, unit: str | None, metrics: dict) -> bool:
    return source == "cma" or unit == "alert" or bool(metrics.get("cma_level") or metrics.get("cma_alertscore"))


def official_tone(headline: str | None, metrics: dict) -> str | None:
    raw = metrics.get("gdacs_alert") or metrics.get("usgs_alert")
    if raw:
        k = str(raw).lower()
        if k in ("green", "yellow", "orange", "red"):
            return k
    lv = cma_level_from_text(f"{headline or ''} {metrics.get('cma_headline') or ''}") or metrics.get("cma_level")
    return _CMA_LEVEL_TONE.get(str(lv)) if lv else None


def _g(band: str, tone: str, zh: str, en: str) -> dict:
    return {"band": band, "tone": tone if tone in TONES else "neutral", "zh": zh, "en": en}


def compute_grade(*, type_: str, magnitude: float | None, unit: str | None,
                  confidence: float | None, source: str | None,
                  headline: str | None, metrics: dict | None) -> dict:
    met = metrics if isinstance(metrics, dict) else {}
    mag = None
    if magnitude is not None:
        try:
            mag = float(magnitude)
        except (TypeError, ValueError):
            mag = None
    tone_off = official_tone(headline, met)
    azh = _ALERT_ZH.get(tone_off or "", "")
    aen = _ALERT_EN.get(tone_off or "", "")
    t = type_

    if t == "earthquake":
        if mag is not None:
            if mag >= 8: band, zh, en = "great", "巨大地震", "Great"
            elif mag >= 7: band, zh, en = "major", "大地震", "Major"
            elif mag >= 6: band, zh, en = "strong", "强震", "Strong"
            elif mag >= 5: band, zh, en = "moderate", "中强震", "Moderate"
            elif mag >= 4: band, zh, en = "light", "有感地震", "Light"
            elif mag >= 3: band, zh, en = "minor", "小震", "Minor"
            else: band, zh, en = "micro", "微震", "Micro"
            tone = "red" if mag >= 7 else "yellow" if mag >= 5 else "neutral"
            pz = f"{azh}警报 · " if tone_off else ""
            pe = f"{aen} alert · " if tone_off else ""
            return _g(band, tone, f"{pz}{zh} M{mag:.1f}", f"{pe}{en} M{mag:.1f}")
        if tone_off:
            return _g("alert", tone_off, f"{azh}警报", f"{aen} alert")
        return _g("na", "neutral", "震级未定", "Magnitude n/a")

    if t == "cyclone":
        if mag is not None:
            if mag >= 137: band, zh, en, tone = "cat5", "五级飓风", "Cat-5 hurricane", "red"
            elif mag >= 113: band, zh, en, tone = "cat4", "四级飓风", "Cat-4 hurricane", "red"
            elif mag >= 96: band, zh, en, tone = "cat3", "三级飓风", "Cat-3 hurricane", "orange"
            elif mag >= 83: band, zh, en, tone = "cat2", "二级飓风", "Cat-2 hurricane", "orange"
            elif mag >= 64: band, zh, en, tone = "cat1", "一级飓风", "Cat-1 hurricane", "yellow"
            elif mag >= 34: band, zh, en, tone = "ts", "热带风暴", "Tropical storm", "yellow"
            else: band, zh, en, tone = "td", "热带低压", "TD", "green"
            pz = f"{azh} · " if tone_off else ""
            pe = f"{aen} · " if tone_off else ""
            return _g(band, tone_off or tone, f"{pz}{zh} {fmt_num(mag)} 节", f"{pe}{en} {fmt_num(mag)} kts")
        if tone_off:
            return _g("alert", tone_off, f"{azh}警报 · 气旋", f"{aen} · cyclone")
        return _g("na", "neutral", "气旋等级未定", "Cyclone grade n/a")

    if t == "wildfire":
        if mag is not None and unit == "MW":
            if mag >= 5000: band, zh, en, tone = "extreme", "特大火点簇", "Extreme fire cluster", "red"
            elif mag >= 1000: band, zh, en, tone = "large", "大型火点簇", "Large fire cluster", "orange"
            elif mag >= 300: band, zh, en, tone = "medium", "中型火点簇", "Medium fire cluster", "yellow"
            elif mag >= 50: band, zh, en, tone = "notable", "较大火点簇", "Notable fire cluster", "yellow"
            else: band, zh, en, tone = "small", "小型火点簇", "Small fire cluster", "green"
            return _g(band, tone, f"{zh} FRP {fmt_num(mag)} MW", f"{en} FRP {fmt_num(mag)} MW")
        if mag is not None:
            if mag >= 100000: band, zh, en, tone = "extreme", "特大火场", "Extremely large fire", "red"
            elif mag >= 10000: band, zh, en, tone = "large", "大型火场", "Large fire", "orange"
            elif mag >= 1000: band, zh, en, tone = "medium", "中型火场", "Medium fire", "yellow"
            elif mag >= 100: band, zh, en, tone = "notable", "较大火场", "Notable fire", "yellow"
            else: band, zh, en, tone = "small", "小型火场", "Small fire", "green"
            pz = f"{azh} · " if tone_off else ""
            pe = f"{aen} · " if tone_off else ""
            return _g(band, tone_off or tone, f"{pz}{zh} {fmt_num(mag)} 英亩", f"{pe}{en} {fmt_num(mag)} acres")
        if tone_off:
            return _g("alert", tone_off, f"{azh}警报 · 野火", f"{aen} alert · wildfire")
        return _g("na", "neutral", "火场规模未定", "Burn area n/a")

    if t == "rainstorm":
        if tone_off:
            return _g("alert", tone_off, f"{azh}暴雨预警", f"{aen} rainstorm warning")
        return _g("alert", "orange", "暴雨预警", "Rainstorm warning")

    if t == "flood":
        if is_cma_alert(source, unit, met) and tone_off:
            return _g("alert", tone_off, f"{azh}山洪/洪水风险预警", f"{aen} flood-risk warning")
        if tone_off:
            if mag is not None and unit != "alert":
                return _g("alert", tone_off, f"{azh}警报 · 径流 {fmt_num(mag)} m³/s",
                          f"{aen} alert · discharge {fmt_num(mag)} m³/s")
            return _g("alert", tone_off, f"{azh}警报 · 洪水", f"{aen} alert · flood")
        if mag is not None and unit != "alert":
            ratio = met.get("ratio")
            try:
                r = float(ratio) if ratio is not None else None
            except (TypeError, ValueError):
                r = None
            if r is not None:
                tone = "red" if r >= 7 else "orange" if r >= 4 else "yellow"
                return _g("discharge", tone, f"径流量 {fmt_num(mag)} m³/s · 基线 {fmt_num(r)} 倍",
                          f"Discharge {fmt_num(mag)} m³/s · {fmt_num(r)}× baseline")
            return _g("discharge", "yellow", f"径流量 {fmt_num(mag)} m³/s", f"Discharge {fmt_num(mag)} m³/s")
        return _g("na", "neutral", "洪水等级未定", "Flood grade n/a")

    if t == "volcano":
        if tone_off:
            return _g("alert", tone_off, f"{azh}警报 · 火山活动", f"{aen} alert · volcano")
        if mag is not None:
            return _g("activity", "orange", f"火山活动 {fmt_num(mag)}", f"Volcano {fmt_num(mag)}")
        return _g("activity", "orange", "火山活动", "Volcano activity")

    if t == "drought":
        if tone_off:
            return _g("alert", tone_off, f"{azh}警报 · 干旱", f"{aen} alert · drought")
        return _g("alert", "yellow", "干旱预警", "Drought alert")

    if t == "war":  # 遗留类型
        level = met.get("war_level") or ("high" if (mag or 0) >= 2.5 else "medium")
        if level == "high":
            return _g("high", "red", "高强度战区 / 持续冲突", "High-intensity war zone / ongoing conflict")
        if level == "medium":
            return _g("medium", "orange", "中等强度冲突关注区", "Medium-intensity conflict zone")
        return _g("low", "yellow", "冲突关注区", "Conflict watch area")

    if t in ("crisis_signal", "armed_clash"):
        conf = float(confidence) if confidence is not None else 0.0
        n = mag if mag is not None else (met.get("article_count") or met.get("event_count") or 0)
        if conf >= 0.7: band, zh, en, tone = "high", "高置信", "High conf.", "red"
        elif conf >= 0.45: band, zh, en, tone = "medium", "中置信", "Medium conf.", "orange"
        elif conf >= 0.25: band, zh, en, tone = "low", "低置信", "Low conf.", "yellow"
        else: band, zh, en, tone = "low", "低置信", "Low conf.", "neutral"
        uz, ue = ("起事件", "events") if unit == "events" else ("篇报道", "articles")
        return _g(band, tone, f"{zh} · {fmt_num(n)} {uz}", f"{en} · {fmt_num(n)} {ue}")

    if tone_off:
        return _g("alert", tone_off, f"{azh}警报", f"{aen} alert")
    if mag is not None:
        u = unit or ""
        return _g("value", "neutral", f"{fmt_num(mag)} {u}".strip(), f"{fmt_num(mag)} {u}".strip())
    return _g("na", "neutral", "等级待定", "Grade n/a")


def grade_for_row(r) -> dict:
    """SQLAlchemy Row（event 表列）→ grade。"""
    return compute_grade(
        type_=r.type, magnitude=r.magnitude_value, unit=r.magnitude_unit,
        confidence=r.confidence, source=r.primary_source, headline=r.headline,
        metrics=r.metrics if isinstance(r.metrics, dict) else None,
    )

File: web/src/grade.js
/** 分级（服务端 grade 优先）、CMA 等级解析、类型/来源/国家标签与颜色（纯函数）。 */
import { EQ_COLOR_M5, EQ_COLOR_M7, TYPE_COLORS, TYPE_COLOR_DEFAULT } from "./constants.js";
import { L, getLang } from "./i18n/index.js";
import { escapeHtml, fmtNum } from "./util/format.js";

/** 地图标记色 = 列表/等级色（地震按震级，其它按类型） */
function mapMarkerColor(p) {
  if (!p) return TYPE_COLOR_DEFAULT;
  if (p.type === "earthquake") return typeTextColor(p);
  return typeColor(p.type);
}

function isCmaAlert(p) {
  if (!p) return false;
  if (p.source === "cma" || p.unit === "alert") return true;
  const met = parseMetrics(p);
  return Boolean(met.cma_level || met.cma_alertscore);
}

function cmaAlertBlob(p) {
  const met = parseMetrics(p);
  return `${p?.headline || ""} ${met.cma_headline || ""}`;
}

function parseCmaLevelFromText(text) {
  const t = String(text || "");
  if (/红色预警|红预警/.test(t)) return "红";
  if (/橙色预警|橙预警/.test(t)) return "橙";
  if (/黄色预警|黄预警/.test(t)) return "黄";
  if (/蓝色预警|蓝预警/.test(t)) return "蓝";
  return null;
}

function cmaSignalLevel(p) {
  return parseCmaLevelFromText(cmaAlertBlob(p)) || parseMetrics(p).cma_level || null;
}

function officialAlertTone(p) {
  const met = parseMetrics(p);
  const raw = met.gdacs_alert || met.usgs_alert;
  if (raw) {
    const k = String(raw).toLowerCase();
    if (["green", "yellow", "orange", "red"].includes(k)) return k;
  }
  const lv = cmaSignalLevel(p);
  if (lv === "红") return "red";
  if (lv === "橙") return "orange";
  if (lv === "黄") return "yellow";
  if (lv === "蓝") return "green";
  return null;
}

function isMinorCmaAlert(p) {
  const lv = cmaSignalLevel(p);
  return lv === "蓝" || lv === "黄";
}

function remapCmaDisplayType(p) {
  if (!p || !isCmaAlert(p)) return p?.type;
  const blob = cmaAlertBlob(p);
  if (/台风|热带风暴|热带低压|风暴潮/.test(blob)) return "cyclone";
  if (/山洪|洪水/.test(blob)) return "flood";
  if (/暴雨|强降雨|地质灾害/.test(blob)) return "rainstorm";
  if (/森林火|草原火/.test(blob)) return "wildfire";
  if (/干旱/.test(blob)) return "drought";
  return p.type;
}

function normalizeEventFeature(f) {
  const p = (f && f.properties) || {};
  const type = remapCmaDisplayType(p) || p.type;
  const lv = cmaSignalLevel(p);
  const met = { ...parseMetrics(p) };
  if (lv && met.cma_level !== lv) met.cma_level = lv;
  return {
    ...f,
    properties: { ...p, type, metrics: met },
  };
}

function typeLabel(t) {
  return L().types[t] || t || L().unknown;
}

function catLabel(c) {
  return L().cats[c] || c || L().unknown;
}

function sourceLabel(s) {
  return L().sources[s] || s || L().unknown;
}

function unitLabel(u) {
  return L().units[u] || u || "";
}

function typeColor(t) {
  return TYPE_COLORS[t] || TYPE_COLOR_DEFAULT;
}

/**
 * 「地震」类型文字颜色（列表 / 弹窗）：
 * 震级 ≥7 → 红；≥5 → 黄；其余用类型默认色
 */
function typeTextColor(p) {
  const base = typeColor(p && p.type);
  if (!p || p.type !== "earthquake") return base;
  const mag =
    p.magnitude != null && p.magnitude !== "" && p.magnitude !== "null" ? Number(p.magnitude) : NaN;
  if (!Number.isFinite(mag)) return base;
  if (mag >= 7) return EQ_COLOR_M7;
  if (mag >= 5) return EQ_COLOR_M5;
  return base;
}

function typeTextStyle(p) {
  const c = typeTextColor(p);
  const bg = c.length === 7 ? `${c}22` : "rgba(240,180,41,0.14)";
  return `color:${c};background:${bg};border-color:${c}55`;
}

const ISO3_ZH = {
  CHN: "中国",
  USA: "美国",
  RUS: "俄罗斯",
  JPN: "日本",
  IND: "印度",
  GBR: "英国",
  FRA: "法国",
  DEU: "德国",
  UKR: "乌克兰",
  IRN: "伊朗",
  ISR: "以色列",
  TWN: "台湾",
  KOR: "韩国",
  PRK: "朝鲜",
};

/** 英文国名（从 place-labels 国家点填充） */
const ISO3_EN = {};

function countryLabel(iso) {
  if (!iso) return L().unknownRegion;
  const code = String(iso).toUpperCase();
  if (getLang() === "en") return ISO3_EN[code] || code;
  return ISO3_ZH[code] || code;
}

/** 加载完整中英地名表（国家中文名） */
async function loadIso3Zh() {
  try {
    const r = await fetch("/data/iso3-zh.json");
    if (!r.ok) return;
    const data = await r.json();
    if (data && typeof data === "object") Object.assign(ISO3_ZH, data);
  } catch (err) {
    console.warn("iso3-zh load", err);
  }
}

function parseMetrics(p) {
  let m = p && p.metrics;
  if (typeof m === "string") {
    try {
      m = JSON.parse(m);
    } catch {
      m = {};
    }
  }
  return m && typeof m === "object" ? m : {};
}

/**
 * 真实灾害等级（来自数据源物理量 / 官方警报，而非仅 0~1 severity）
 * 返回 { text, tone } tone: green|yellow|orange|red|neutral
 */
function realGrade(p) {
  // 服务端已统一分级（/api/events grade 字段）；本地实现仅作夹具/旧数据兜底
  const g = p && p.grade;
  if (g && typeof g === "object" && g.tone) {
    return { text: getLang() === "en" ? g.en || g.zh || "" : g.zh || g.en || "", tone: g.tone };
  }
  const t = p.type;
  const met = parseMetrics(p);
  const mag =
    p.magnitude != null && p.magnitude !== "" && p.magnitude !== "null"
      ? Number(p.magnitude)
      : null;
  const alertTone = officialAlertTone(p);
  const alertKey = alertTone || "";

  // 官方警报色（GDACS / USGS / 中央气象台）
  const en = getLang() === "en";
  const alertZh = en
    ? { green: "Green", yellow: "Yellow", orange: "Orange", red: "Red" }
    : { green: "绿色", yellow: "黄色", orange: "橙色", red: "红色" };

  if (t === "earthquake") {
    if (mag != null && Number.isFinite(mag)) {
      let band = en ? "Micro" : "微震";
      if (mag >= 8) band = en ? "Great" : "巨大地震";
      else if (mag >= 7) band = en ? "Major" : "大地震";
      else if (mag >= 6) band = en ? "Strong" : "强震";
      else if (mag >= 5) band = en ? "Moderate" : "中强震";
      else if (mag >= 4) band = en ? "Light" : "有感地震";
      else if (mag >= 3) band = en ? "Minor" : "小震";
      const tone = mag >= 7 ? "red" : mag >= 5 ? "yellow" : "neutral";
      const alertPart = alertTone
        ? en
          ? `${alertZh[alertKey]} alert · `
          : `${alertZh[alertKey]}警报 · `
        : "";
      return { text: `${alertPart}${band} M${mag.toFixed(1)}`, tone };
    }
    if (alertTone)
      return {
        text: en ? `${alertZh[alertKey]} alert` : `${alertZh[alertKey]}警报`,
        tone: alertTone,
      };
    return { text: en ? "Magnitude n/a" : "震级未定", tone: "neutral" };
  }

  if (t === "cyclone") {
    if (mag != null && Number.isFinite(mag)) {
      let band = en ? "TD" : "热带低压";
      let tone = "green";
      if (mag >= 137) {
        band = en ? "Cat-5 hurricane" : "五级飓风";
        tone = "red";
      } else if (mag >= 113) {
        band = en ? "Cat-4 hurricane" : "四级飓风";
        tone = "red";
      } else if (mag >= 96) {
        band = en ? "Cat-3 hurricane" : "三级飓风";
        tone = "orange";
      } else if (mag >= 83) {
        band = en ? "Cat-2 hurricane" : "二级飓风";
        tone = "orange";
      } else if (mag >= 64) {
        band = en ? "Cat-1 hurricane" : "一级飓风";
        tone = "yellow";
      } else if (mag >= 34) {
        band = en ? "Tropical storm" : "热带风暴";
        tone = "yellow";
      }
      const alertPart = alertTone ? `${alertZh[alertKey]} · ` : "";
      return {
        text: `${alertPart}${band} ${fmtNum(mag)} ${en ? "kts" : "节"}`,
        tone: alertTone || tone,
      };
    }
    if (alertTone)
      return {
        text: en ? `${alertZh[alertKey]} · cyclone` : `${alertZh[alertKey]}警报 · 气旋`,
        tone: alertTone,
      };
    return { text: en ? "Cyclone grade n/a" : "气旋等级未定", tone: "neutral" };
  }

  if (t === "wildfire") {
    if (mag != null && Number.isFinite(mag)) {
      // 过火面积（英亩）官方数据分档
      let band = en ? "Small fire" : "小型火场";
      let tone = "green";
      if (mag >= 100000) {
        band = en ? "Extremely large fire" : "特大火场";
        tone = "red";
      } else if (mag >= 10000) {
        band = en ? "Large fire" : "大型火场";
        tone = "orange";
      } else if (mag >= 1000) {
        band = en ? "Medium fire" : "中型火场";
        tone = "yellow";
      } else if (mag >= 100) {
        band = en ? "Notable fire" : "较大火场";
        tone = "yellow";
      }
      const alertPart = alertTone ? `${alertZh[alertKey]} · ` : "";
      return {
        text: `${alertPart}${band} ${fmtNum(mag)} ${en ? "acres" : "英亩"}`,
        tone: alertTone || tone,
      };
    }
    if (alertTone)
      return {
        text: en ? `${alertZh[alertKey]} alert · wildfire` : `${alertZh[alertKey]}警报 · 野火`,
        tone: alertTone,
      };
    return { text: en ? "Burn area n/a" : "火场规模未定", tone: "neutral" };
  }

  if (t === "rainstorm") {
    if (alertTone) {
      return {
        text: en ? `${alertZh[alertKey]} rainstorm warning` : `${alertZh[alertKey]}暴雨预警`,
        tone: alertTone,
      };
    }
    return { text: en ? "Rainstorm warning" : "暴雨预警", tone: "orange" };
  }

  if (t === "flood") {
    if (isCmaAlert(p) && alertTone) {
      return {
        text: en
          ? `${alertZh[alertKey]} flood-risk warning`
          : `${alertZh[alertKey]}山洪/洪水风险预警`,
        tone: alertTone,
      };
    }
    if (alertTone) {
      return {
        text:
          mag != null && Number.isFinite(mag) && p.unit !== "alert"
            ? en
              ? `${alertZh[alertKey]} alert · discharge ${fmtNum(mag)} m³/s`
              : `${alertZh[alertKey]}警报 · 径流 ${fmtNum(mag)} m³/s`
            : en
              ? `${alertZh[alertKey]} alert · flood`
              : `${alertZh[alertKey]}警报 · 洪水`,
        tone: alertTone,
      };
    }
    if (mag != null && Number.isFinite(mag) && p.unit !== "alert") {
      return {
        text: en ? `Discharge ${fmtNum(mag)} m³/s` : `径流量 ${fmtNum(mag)} m³/s`,
        tone: "yellow",
      };
    }
    return { text: en ? "Flood grade n/a" : "洪水等级未定", tone: "neutral" };
  }

  if (t === "volcano") {
    if (alertTone)
      return {
        text: en ? `${alertZh[alertKey]} alert · volcano` : `${alertZh[alertKey]}警报 · 火山活动`,
        tone: alertTone,
      };
    return {
      text:
        mag != null
          ? en
            ? `Volcano ${fmtNum(mag)}`
            : `火山活动 ${fmtNum(mag)}`
          : en
            ? "Volcano activity"
            : "火山活动",
      tone: "orange",
    };
  }

  if (t === "drought") {
    if (alertTone)
      return {
        text: en ? `${alertZh[alertKey]} alert · drought` : `${alertZh[alertKey]}警报 · 干旱`,
        tone: alertTone,
      };
    return { text: en ? "Drought alert" : "干旱预警", tone: "yellow" };
  }

  if (t === "war") {
    const level = met.war_level || (mag != null && mag >= 2.5 ? "high" : "medium");
    if (level === "high")
      return {
        text: en ? "High-intensity war zone / ongoing conflict" : "高强度战区 / 持续冲突",
        tone: "red",
      };
    if (level === "medium")
      return { text: en ? "Medium-intensity conflict zone" : "中等强度冲突关注区", tone: "orange" };
    return { text: en ? "Conflict watch area" : "冲突关注区", tone: "yellow" };
  }

  if (t === "crisis_signal" || t === "armed_clash") {
    const conf = Number(p.confidence);
    const arts = Number.isFinite(mag) ? mag : met.article_count || met.event_count || 0;
    let band = en ? "Low conf." : "低置信";
    let tone = "neutral";
    if (conf >= 0.7) {
      band = en ? "High conf." : "高置信";
      tone = "red";
    } else if (conf >= 0.45) {
      band = en ? "Medium conf." : "中置信";
      tone = "orange";
    } else if (conf >= 0.25) {
      band = en ? "Low conf." : "低置信";
      tone = "yellow";
    }
    const unit = p.unit === "events" ? (en ? "events" : "起事件") : en ? "articles" : "篇报道";
    return { text: `${band} · ${fmtNum(arts)} ${unit}`, tone };
  }

  if (alertTone)
    return {
      text: en ? `${alertZh[alertKey]} alert` : `${alertZh[alertKey]}警报`,
      tone: alertTone,
    };
  if (mag != null && Number.isFinite(mag)) {
    return { text: `${fmtNum(mag)} ${unitLabel(p.unit) || ""}`.trim(), tone: "neutral" };
  }
  return { text: en ? "Grade n/a" : "等级待定", tone: "neutral" };
}

function gradeBadgeHtml(p) {
  const g = realGrade(p);
  return `<span class="grade-badge g-${g.tone}">${escapeHtml(g.text)}</span>`;
}

export {
  ISO3_ZH,
  ISO3_EN,
  countryLabel,
  loadIso3Zh,
  parseMetrics,
  realGrade,
  gradeBadgeHtml,
  officialAlertTone,
  isCmaAlert,
  cmaAlertBlob,
  parseCmaLevelFromText,
  cmaSignalLevel,
  isMinorCmaAlert,
  remapCmaDisplayType,
  normalizeEventFeature,
  typeLabel,
  catLabel,
  sourceLabel,
  unitLabel,
  typeColor,
  typeTextColor,
  typeTextStyle,
  mapMarkerColor,
};
